File: js_practice/heap_sort.js
const _ = require("lodash");


console.log(_.empty([]));
debugger;

function heapSort(arr) {
  let heap = [];

  for (let i = 0; i < arr.length; i++) {
    heap.push(arr[i]);
    heap.heapifyUp(heap.length-1);
  }

  let sortedArr = [];
  while (heap.length > 0) {
    [ heap[0], heap[heap.length - 1] ] = [ heap[heap.length - 1], heap[0] ];
    sortedArr.push(heap.pop());
    heap.heapifyDown(0);
  }
  return sortedArr;
}


Array.prototype.heapifyUp = function(idx) {
  let childIdx = idx;
  let complete = false;
  while (!complete && childIdx > 0) {
    complete = true;
    console.log(childIdx);
    let parent = parentIdx(childIdx);
    if (this[parent] > this[childIdx]) {
      [ this[parent], this[childIdx] ] = [ this[childIdx], this[parent] ]
      complete = false;
      childIdx = parentIdx;
    }
  }
  // console.log(this);
}

[1,2,3,4,5,6,7,8,9,10,2.5].heapifyUp(10);

Array.prototype.heapifyDown = function(idx) {
  let parentIdx = idx;
  let children = childIdx(idx);

  // if (children.empty())
}


function parentIdx(childIdx) {
  return childIdx === 0 ? undefined : Math.floor((childIdx - 1) / 2)
}

function childIdx(parentIdx, length) {
  let children = [];
  let child1 = (2 * parentIdx) + 1;
  let child2 = (2 * parentIdx) + 2;
  if (child1 < length) { children.push(child1) }
  if (child2 < length) { children.push(child2) }
  return children;
}

File: algorithms/heapsort.rb
require 'byebug'

def heapsort(arr)
  # Place all elements of array into a heap
  # Heapify down
  heap = []

  arr.each do |el|
    heap.push(el)
    heapify_up(heap, heap.length-1)
    p heap
  end
  p "final heap:"
  p heap
  # Remove all elements of heap into array
  # Swap first/last & heapify up
  sorted = []
  until heap.empty?
    heap[0], heap[-1] = heap[-1], heap[0]
    el = heap.pop()
    sorted.push(el)
    heapify_down(heap, 0)
    p heap
  end

  p "final sorted:"
  p sorted
end


def heapify_up(heap, idx)
  return if idx == 0
  child_idx = idx
  parent_idx = (idx - 1) / 2

  until child_idx == 0 || heap[parent_idx] <= heap[child_idx]
    heap[child_idx], heap[parent_idx] = heap[parent_idx], heap[child_idx]
    child_idx = parent_idx
    parent_idx = (idx - 1) / 2
  end

  heap
end

def heapify_down(heap, idx)
  parent_idx = idx
  child_idxs = children(idx, heap.length)
  return if child_idxs.empty?

  child_idx = child_idxs.min_by{ |el| heap[el] }
  p child_idxs
  p child_idx

  until child_idx.nil? || parent_idx >= heap.length || heap[parent_idx] <= heap[child_idx]
    heap[parent_idx], heap[child_idx] = heap[child_idx], heap[parent_idx]
    parent_idx = child_idx
    child_idxs = children(idx, heap.length)
    child_idx = child_idxs.min_by{ |el| heap[el] }
  end

  heap
end

def children(idx, length)
  child_idxs = [idx * 2 + 1, idx * 2 + 2]
  child_idxs.select { |el| el < length }
end

# p heapsort([7,1,3,4])
# p heapsort([1,4,0,5,2])
p heapsort([3,4,0,8,7,2,7,9])
# p heapsort([3,-4,0,8,-7,2,-4,7,-9])
# p heapsort([0,0,0,0,0])
# p heapsort([9,7,6,4,3,2,1,-5])
